File: src/main/java/net/zenxarch/bot/defense/modules/WaterMLG.java
package net.zenxarch.bot.defense.modules;

import static net.zenxarch.bot.util.ClientPlayerHelper.*;

import net.minecraft.block.Blocks;
import net.minecraft.item.Items;
import net.minecraft.util.Hand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.RaycastContext.FluidHandling;
import net.minecraft.world.World;
import net.zenxarch.bot.defense.DefenseStateManager;
import net.zenxarch.bot.util.BlockPlacementUtils;

public class WaterMLG extends Module {
  public WaterMLG() { super("WaterMLG"); }

  @Override
  public void preTick() {
    var saveItemSlot = findSaveItem();
    if (saveItemSlot == -1)
      return;

    if (mc.player.isOnGround() || mc.player.inPowderSnow ||
        mc.player.isTouchingWater())
      return;

    var blocks = getBlocksUntilLanding();
    if (mc.player.fallDistance + blocks < 3)
      return;

    var pos = mc.player.getBlockPos().down(blocks - 1);
    var hit =
        BlockPlacementUtils.raycastToBlockForPlacement(pos, FluidHandling.NONE);
    if (hit == null)
      return;
    DefenseStateManager.performAction(() -> {
      pickItemSlot(saveItemSlot);
      return BlockPlacementUtils.place(hit, Hand.MAIN_HAND);
    });
  }

  private int getBlocksUntilLanding() {
    var start = mc.player.getBlockY();
    var end = start - Math.min(start - mc.world.getBottomY(), 10);
    int i;
    for (i = mc.player.getBlockY(); i > end; i--) {
      var pos = new BlockPos(mc.player.getBlockX(), i, mc.player.getBlockZ());
      if (checkAir(pos))
        continue;
      if (safeToLand(pos))
        return -1;
      break;
    }
    return start - i;
  }

  private boolean checkAir(BlockPos pos) {
    var state = mc.world.getBlockState(pos);
    return state.getCollisionShape(mc.world, pos).isEmpty();
  }

  private boolean safeToLand(BlockPos pos) {
    return !mc.world.getFluidState(pos).isEmpty() ||
        mc.world.getBlockState(pos).isOf(Blocks.POWDER_SNOW);
  }

  private int findSaveItem() {
    var water = findInInventory(Items.WATER_BUCKET);
    if (water != -1 && mc.player.world.getRegistryKey() != World.NETHER) {
      return water;
    }
    return findInInventory(Items.POWDER_SNOW_BUCKET);
  }
}

File: src/main/scala/net/zenxarch/bot/defense/modules/WaterMLG.scala
package net.zenxarch.bot.defense.modules

import net.minecraft.block.Blocks
import net.minecraft.block.FluidDrainable
import net.minecraft.block.FluidFillable
import net.minecraft.client.option.KeyBinding
import net.minecraft.client.util.InputUtil
import net.minecraft.item.BlockItem
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.util.Hand
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.hit.HitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.world.RaycastContext.FluidHandling
import net.minecraft.world.World
import net.zenxarch.bot.defense.DefenseStateManager
import net.zenxarch.bot.util.BlockPlacementUtils._
import net.zenxarch.bot.util.ClientPlayerHelper
import net.zenxarch.bot.util.ClientPlayerHelper._

class WaterMLG extends Module("WaterMlg"):
  import Module.mc

  override def preTick() =
    val saveItemSlot = findSaveItem()
    if saveItemSlot == -1 then return

    if mc.player.isOnGround() || mc.player.inPowderSnow ||
      mc.player.isTouchingWater()
    then return

    var landPos = getLandingBlock(6)
    if landPos == null then return
    val blocks = mc.player.getPos.y.toFloat - landPos.getY.toFloat
    if mc.player.fallDistance + blocks < 4 then return

    landPos = landPos.up
    val side = getPlaceableSide(landPos, checkSide(landPos, _))
    if side == null then return
    val vec = getVecForBlockPlacement(landPos, side)

    val hit = praycast(vec, FluidHandling.NONE)
    if hit == null || !hit.getType().equals(HitResult.Type.BLOCK) then return
    val bhit = hit.asInstanceOf[BlockHitResult]
    if !bhit.getBlockPos().equals(landPos.offset(side)) || !bhit
        .getSide()
        .equals(side.getOpposite)
    then return
    DefenseStateManager.performAction(() => {
      pickItemSlot(saveItemSlot)
      ClientPlayerHelper.lookAt(vec.x, vec.y, vec.z)
      ClientPlayerHelper.syncRotation()
      KeyBinding.onKeyPressed(InputUtil.fromTranslationKey("key.use"))
      true
    })

  private def checkSide(pos: BlockPos, side: Direction): Boolean =
    val block = mc.world.getBlockState(pos.offset(side))
    side != Direction.UP && !block.getBlock().isInstanceOf[FluidFillable] && !block.getBlock().isInstanceOf[FluidDrainable] && block.isSideSolidFullSquare(mc.world, pos.offset(side),side.getOpposite())

  private def getLandingBlock(upto: Int): BlockPos =
    val start = mc.player.getBlockPos()
    val end = Math.min(start.getY() - mc.world.getBottomY(), upto)
    val bb = mc.player.getBoundingBox()
    for
      y <- 1 until end
      x <- bb.minX.toInt to bb.maxX.toInt
      z <- bb.minZ.toInt to bb.maxZ.toInt
    do
      val pos = BlockPos(x, start.getY - y, z)
      if !checkAir(pos) then
        return if safeToLand(pos) then null
        else pos
    return null

  private def checkAir(pos: BlockPos) =
    mc.world.getBlockState(pos).getCollisionShape(mc.world, pos).isEmpty()

  private def safeToLand(pos: BlockPos) =
    !mc.world.getFluidState(pos).isEmpty() ||
      mc.world.getBlockState(pos).isOf(Blocks.POWDER_SNOW)

  private def findSaveItem(): Int =
    var water = findInInventory(Items.WATER_BUCKET)
    if water != -1 && !mc.player.world.getRegistryKey().equals(World.NETHER)
    then return water
    var powder = findInInventory(Items.POWDER_SNOW_BUCKET)
    if powder != -1 then return powder

    return findInInventory(is =>
      !is.isEmpty() && is.getItem().isInstanceOf[BlockItem]
    )
